docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,242 @@
+---
+title: Opt-in Autoscaling from Zero
+authors:
+  - "@elmiko"
+reviewers:
+  - "@"
+creation-date: 2021-03-10
+last-updated: 2021-03-10
+status: implementable
+---
+
+# Opt-in Autoscaling from Zero
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals/Future Work](#non-goalsfuture-work)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+* **Node Group** This term has special meaning within the cluster autoscaler, it refers to collections
+  of nodes, and related physical hardware, that are organized within the autoscaler for scaling operations.
+  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled
+  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond
```

Comment:
IIRC, currently there is no functional difference between a Cluster Autoscaler at a management level and a workload cluster level apart from some basic configuration (please correct me if this is incorrect).
Now this introduces a feature which is available only in the capi world, and non-capi clusters will not have this functionality directly. Was there a reason why this was not done as a cluster-autoscaler feature and more as a capi feature?

Reply:
_some_ of the other providers support scale from zero (i think AWS and GCP), but this is a feature which is available in the upstream for any provider who wishes to implement. it is part of the [node group interface](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L172) and is optional to implement.

> Was there a reason why this was not done as a cluster-autoscaler feature and more as a capi feature?

yes, essentially it must be done by each provider. the advantage for CAPI is that through this provider you will get scale from zero on every platform that CAPI supports. adding this functionality to the other providers in the autoscaler would be nice, but it is a little beyond scope for this effort.